### ISSUE TITLE
Missing NumPy functions needed for SLAC applications

### ISSUE THREAD

**Issue by syamajala** [nv-legate/cunumeric#1116]
I'm opening this issue so @manopapad and I can keep track of what needs to be implemented for the different cunumeric SLAC applications.

For psana we need: numpy.ma.median, np.unique(return_index=True), @manopapad has a patch that tries to improve single index accesses to arrays although that code will be removed when np.uinque(return_index=True) is implemented. All the kernels for psana are just single GPU and do not need to be distributed.

For HDF5 analysis we need gpu and distributed versions of: np.nanquantile, np.nanpercentile, np.logical_and.reduce, np.diff, np.digitize, np.average, np.cov.

**Comment by rohany** [comment]
I have a pending PR for np.diff against cuNumeric that can be dusted off and merged.

**Comment by syamajala** [comment]
We need scipy.optimize.curve_fit. Under the hood this seems to use minpack. Depending on what options you pass to curve_fit I think it might also need cholesky.

**Comment by JosephGuman** [comment]
At present it seems like DeferredArray's unary_reduction() implementation [doesn't allow reducing over multiple dimensions](https://github.com/nv-legate/cunumeric/blob/branch-24.01/cunumeric/deferred.py#L3259), which would probably be needed to average over an arbitrary subset of axes. Is this important to address in this issue, or is it not required for SLAC's application?

**Comment by syamajala** [comment]
We have a need for scipy.curve_fit.

**Comment by syamajala** [comment]
Ok. Will give them a try early next week.

**Comment by JosephGuman** [comment]
I might took a look at np.unique(return_index=True) if nobody else is working on it right now.

**Comment by syamajala** [comment]
We still need to investigate performance issues related to the functions that were implemented in this ticket. 

Here is a profile from before the missing functions were implemented:
https://legion.stanford.edu/prof-viewer/?url=https://sapling.stanford.edu/~seshu/xpp/legion_prof/

And a profile from after:
https://legion.stanford.edu/prof-viewer/?url=https://sapling.stanford.edu/~seshu/xpp/legion_prof.1/

**Comment by rohany** [comment]
There might be still some missing functions that correspond to the pieces of high python utilization, but i don't really see a performance issue in this profile other than the problem size is too small (especially for the public Python core).

**Comment by syamajala** [comment]
The following functions are missing:

- [ ] np.inv
- [ ] np.correlate
- [ ] np.fft.fftshift
- [ ] np.round
- [ ] np.unravel_index
- [ ] np.meshgrid

**Comment by qldnfox** [comment]
Also missing the following:

-  np.fft.ifft

- np.rot90

**Comment by syamajala** [comment]
nansum does not support reducing over multiple dimensions:
```
   File "/sdf/group/lcls/ds/tools/conda_envs/cunumeric-mec/lib/python3.12/site-packages/cunumeric/_module/math_sum_prod_diff.py", line 951, in nansum
    return a._nansum(
           ^^^^^^^^^^
  File "/sdf/group/lcls/ds/tools/conda_envs/cunumeric-mec/lib/python3.12/site-packages/cunumeric/_array/array.py", line 3580, in _nansum
    return a._nansum(
           ^^^^^^^^^^
  File "/sdf/group/lcls/ds/tools/conda_envs/cunumeric-mec/lib/python3.12/site-packages/cunumeric/_array/array.py", line 3580, in _nansum
    return perform_unary_reduction(
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sdf/group/lcls/ds/tools/conda_envs/cunumeric-mec/lib/python3.12/site-packages/cunumeric/_array/thunk.py", line 233, in perform_unary_reduction
    return perform_unary_reduction(
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/sdf/group/lcls/ds/tools/conda_envs/cunumeric-mec/lib/python3.12/site-packages/cunumeric/_array/thunk.py", line
 233, in perform_unary_reduction
    result._thunk.unary_reduction(
  File "/sdf/group/lcls/ds/tools/conda_envs/cunumeric-mec/lib/python3.12/site-packages/cunumeric/_thunk/deferred.py", l
ine 148, in wrapper
    result._thunk.unary_reduction(
  File "/sdf/group/lcls/ds/tools/conda_envs/cunumeric-mec/lib/python3.12/site-packages/cunumeric/_thunk/deferred.py", l
ine 148, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/sdf/group/lcls/ds/tools/conda_envs/cunumeric-mec/lib/python3.12/site-packages/cunumeric/_thunk/deferred.py", l
ine 3192, in unary_reduction
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/sdf/group/lcls/ds/tools/conda_envs/cunumeric-mec/lib/python3.12/site-packages/cunumeric/_thunk/deferred.py", l
ine 3192, in unary_reduction
    raise NotImplementedError(
NotImplementedError: Need support for reducing multiple dimensions
    raise NotImplementedError(
NotImplementedError: Need support for reducing multiple dimensions
```

Also for some reason nanpercentile is still falling back to numpy in cunumeric 24.06.00. It looks like it was merged above though?

**Comment by syamajala** [comment]
Adding these to the list:
- [ ] histogram2d
- [ ] median

**Comment by qldnfox** [comment]
@manopapad is np.meshgrid implemented?

**Comment by manopapad** [comment]
Yes, it's in there https://docs.nvidia.com/cupynumeric/25.03/api/generated/cupynumeric.meshgrid.html

**Comment by sbuneti** [comment]
The MVP has been delivered, agreed to close.
